### ISSUE TITLE
Terraform AWS Provider Version 2.45.0: Missing Error Messages and Not Retrying

### ISSUE THREAD

**Issue by bflad** [terraform-providers/terraform-provider-aws#11682]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform AWS Provider version 2.45.0

### Affected Resource(s)

- `aws_cloud9_environment_ec2`
- `aws_cloudtrail`
- `aws_cloudwatch_log_destination`
- `aws_cloudwatch_log_subscription`
- `aws_codebuild_project`
- `aws_codedeploy_deployment_group`
- `aws_config_delivery_channel`
- `aws_ecs_service`
- `aws_elasticsearch_domain`

Additional resources and errors may be missing error messaging as well, however these are directly affected by resource logic attempting to retry on error messages that are no longer present in the AWS Go SDK as of version 1.28.0.

### Terraform Configuration Files

```hcl
# Resource dependent
```

### Debug Output

Example debug output (note lack of error message after error code on last line):

```
2020/01/20 12:59:15 [DEBUG] [aws-sdk-go] DEBUG: Response codebuild/CreateProject Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 400 Bad Request
Connection: close
Content-Length: 167
Content-Type: application/x-amz-json-1.1
Date: Mon, 20 Jan 2020 17:59:14 GMT
X-Amzn-Requestid: 2f589026-d531-458d-b851-86bd4afb5cd0


-----------------------------------------------------
2020/01/20 12:59:15 [DEBUG] [aws-sdk-go] {"__type":"InvalidInputException","message":"CodeBuild is not authorized to perform: sts:AssumeRole on arn:aws:iam::--OMITTED--:role/tf-acc-test-4387558838057379294"}
2020/01/20 12:59:15 [DEBUG] [aws-sdk-go] DEBUG: Validate Response codebuild/CreateProject failed, attempt 0/25, error InvalidInputException:
```

### Expected Behavior

In affected resources, Terraform AWS Provider outputs error messages and retries on certain errors. e.g. those resources affected by new lack of ability to retry properly:

- In `aws_cloud9_environment_ec2` resource, unable to retry for IAM eventual consistency on `NotFoundException` during creation (https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_cloud9_environment_ec2.go#L96-L98)
- In `aws_cloudtrail` resource, unable to retry for CloudWatch Logs eventual consistency on `InvalidCloudWatchLogsRoleArnException` and `InvalidCloudWatchLogsLogGroupArnException` during creation or update (https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_cloudtrail.go#L186-L191)
- In `aws_cloudwatch_log_destination` resource, unable to retry for `InvalidParameterException` with test message delivery during creation or update (https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_cloudwatch_log_destination.go#L70-L72)
- In `aws_cloudwatch_log_subscription` resource, unable to retry for `InvalidParameterException` with test message delivery and Lambda Function execution during creation (https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_cloudwatch_log_subscription_filter.go#L71-L76)
- In `aws_codebuild_project` resource, unable to retry for IAM eventual consistency on `InvalidInputException` during creation and update (https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_codebuild_project.go#L665-L669)
- In `aws_codedeploy_deployment_group` resource, unable to retry for SNS Topic eventual consistency on `InvalidTriggerConfigException` during creation and update (https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_codedeploy_deployment_group.go#L783-L788)
- In `aws_config_delivery_channel` resource, unable to retry for running Config Configuration Recorder on `LastDeliveryChannelDeleteFailedException` during deletion (https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_config_delivery_channel.go#L172-L174)
- In `aws_ecs_service` resource, unable to retry for cross-service eventual consistency on `InvalidParameterException` during creation and update (https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_ecs_service.go#L524-L533)
- In `aws_elasticsearch_domain` resource, unable to retry for various eventual consistency errors on `InvalidTypeException` / `ValidationException` during creation (https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_elasticsearch_domain.go#L459-L474)

Historically, we have preferred to check specific error messaging from the service API responses since the error code itself is not descriptive enough to determine if the error could/should be retried. While the service APIs should likely introduce more granular error codes and fix their API models, the unexpected removal of the error messaging represents a fairly problematic change in the AWS Go SDK since it was previously supported by many services.

### Actual Behavior

Error messages (after the error code and colon when printed) are missing (an empty string). Retries in certain resources are not performed as expected since the error message conditions are no longer true.

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

Changes in v1.28.0 of the AWS Go SDK (imported and released by Terraform AWS Provider version 2.45.0) introduced code generated types for certain service error shapes, which can be used to extract additional information provided by the API response. Prior to these code generated types, most (if not all) of these errors returned the correct error messaging via `Message()` via the `message` protocol response field. However after this change, since some of these error shapes in the service API models are incorrectly missing the `Message` structure member, this causes the `Message_` (Go struct type) to be missing, which causes the error message via `Message()` to always be empty.

An example of this missing error messaging can be seen with `codebuild` service `InvalidInputException` shape:

https://github.com/aws/aws-sdk-go/blob/6627e8938cdef5476dec0510b524258d3bf3983e/models/apis/codebuild/2016-10-06/api-2.json#L998-L1003

https://github.com/aws/aws-sdk-go/blob/6627e8938cdef5476dec0510b524258d3bf3983e/service/codebuild/api.go#L5333-L5336

Full SDK analysis of this new behavior was performed here: https://github.com/terraform-providers/terraform-provider-aws/issues/11642#issuecomment-575820896

### References

- https://github.com/aws/aws-sdk-go/issues/3088
- https://github.com/aws/aws-sdk-go/pull/3062
- https://github.com/terraform-providers/terraform-provider-aws/pull/11501
- https://github.com/terraform-providers/terraform-provider-aws/issues/11642
- https://github.com/terraform-providers/terraform-provider-aws/issues/7725

**Comment by bflad** [comment]
Created upstream AWS Go SDK issue: https://github.com/aws/aws-sdk-go/issues/3088

**Comment by bflad** [comment]
Upstream fix https://github.com/aws/aws-sdk-go/pull/3089 appears to resolve this issue -- we'll get this dependency update merged as soon as its available upstream.

**Comment by bflad** [comment]
The upstream fix for this has been merged and should go out with today's AWS Go SDK release.

**Comment by bflad** [comment]
Oddly enough, there was not an AWS Go SDK release on its normal daily release cadence today. To ensure we have fully run the provider test suite (occurs nightly), we have merged in the expected fix directly. If there are no additional regressions, this fix will release tomorrow with version 2.46.0 of the Terraform AWS Provider.
